### PR TITLE
[WIP][FEATURE] added handler for unsaved changes

### PR DESCRIPTION
**Pull request by michaelzangerle**
__Tasks:__
- [x] implement overlay for unsaved changes
- [ ] special cases for new entities (description below)
- [ ] events should be renamed because of name collision in content bundle
- [ ] should be integrated in the content bundle
- [ ] got feedback

__Informations:__

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Tests pass?   | none
| Fixed tickets | none
| Doc           | none

__Requires:__
- https://github.com/massiveart/husky/pull/443

__Colliding navigate events__
When new entity is created the user will be redirected from .../add to .../entity/[id]. This collides with the handling of unsaved changes in one case: When a new entity is created (not saved) and the navigates, he will be asked whether he want to save the changes or not. When he saves the changes the unsaved change handling will navigate to wherever he clicked but the existing logic gets triggered later and the user returns to .../entity/[id]!

### CODE REVIEW

File: src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js [hunk 1]
```diff
@@ -145,6 +147,63 @@ define(function() {
...
 
         /**
+        * Initializes the handler for unsaved changes
+        */
+        initializeUnsavedChangesHandler: function(){
+            this.unsavedChanges = false;
+            this.bindCustomEventsForUnsavedChangesHandler();
+        },
+
+        /**
```
Comment (dantleech) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js]: indentiation pb.

File: src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js [hunk 2]
```diff
@@ -145,6 +147,63 @@ define(function() {
...
+            }.bind(this));
+            this.sandbox.on('husky.toolbar.header.item.disable', function(button) {
+                if (button === constants.saveButtonKey) {
+                    this.unsavedChanges = false;
+                }
+            }.bind(this));
+        },
+
+        /**
+        * Checks if there are unsaved changes and shows a message box
```
Comment (dantleech) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js]: didob

File: src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js [hunk 3]
```diff
@@ -145,6 +147,72 @@ define(function() {
...
+        initializeUnsavedChangesHandler: function() {
+            this.unsavedChanges = false;
+            this.bindCustomEventsForUnsavedChangesHandler();
+        },
+
+        /**
+         * Listen for changes regarding the save button
+         */
+        bindCustomEventsForUnsavedChangesHandler: function() {
+            this.sandbox.on('husky.toolbar.header.item.enable', function(button) {
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js]: shouldn´t we introduce a content-changes event? i think thats a bit hacky?

File: src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js [hunk 4]
```diff
@@ -249,6 +317,7 @@ define(function() {
          * @param forceReload {Boolean} force page to reload
          */
         navigate: function(route, trigger, noLoader, forceReload) {
+
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js]: why did you add empty lines here ...

File: src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js [hunk 5]
```diff
@@ -264,6 +333,7 @@ define(function() {
             // navigate
             router.navigate(route, {trigger: trigger});
             this.sandbox.dom.scrollTop(this.sandbox.dom.$window, 0);
+
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js]: and here?

File: src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js [hunk 6]
```diff
@@ -289,7 +359,28 @@ define(function() {
          */
         bindCustomEvents: function() {
             // navigate
-            this.sandbox.on('sulu.router.navigate', this.navigate.bind(this));
+            this.sandbox.on('sulu.router.navigate', function(route, trigger, noLoader, forceReload) {
+                if (!!this.unsavedChanges) {
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js]: move this to a `preNavigate` function!

File: src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js (hunk 3 above)
Comment (michaelzangerle) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js]: we probably should have introduced one at the start but I didn't want to change every bundle!

File: src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js [hunk 7]
```diff
@@ -289,7 +359,28 @@ define(function() {
...
+                        this.navigate(route, trigger, noLoader, forceReload);
+                    }.bind(this));
+                } else {
+                    this.navigate(route, trigger, noLoader, forceReload);
+                }
+            }.bind(this));
+
+            // navigate when tab element of header is clicked
+            this.sandbox.on('husky.tabs.header.item.preselect', function(event) {
+                if (!!this.unsavedChanges) {
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js]: move this to a `preSelectTab` function!

File: src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js [hunk 8]
```diff
@@ -289,7 +359,28 @@ define(function() {
...
+                    var dfdSavedChanges = this.showUnsavedChangesWarning();
+                    this.sandbox.data.when(dfdSavedChanges).then(function() {
+                        this.sandbox.emit('husky.tabs.header.item.clicked', event);
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js]: thats not husky ... please throw a event with a suitable name!

File: src/Sulu/Bundle/AdminBundle/Resources/public/js/components/header/main.js [hunk 9]
```diff
@@ -54,7 +54,12 @@ define([], function () {
             contentEl: null,
             toolbarOptions: {},
             toolbarLanguageChanger: true,
-            tabsOptions: {},
+            tabsOptions: {
+                preSelectItem: {
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/components/header/main.js]: use `preSelectEvent`

File: src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js (hunk 8 above)
Comment (michaelzangerle) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js]: I want to trigger an event at the husky component?

File: src/Sulu/Bundle/AdminBundle/Resources/translations/sulu/backend.en.xlf [hunk 10]
```diff
@@ -1434,6 +1434,22 @@
                 <source>labels.successfully-saved</source>
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/AdminBundle/Resources/translations/sulu/backend.en.xlf]: should have the same id in english and german `a504f522de1456524142c0b4820e29a5f `

File: src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js (hunk 3 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js]: but that should be done! what did you think @chirimoya @danrot @dantleech

Comment (dantleech) [inline src/Sulu/Bundle/AdminBundle/Resources/public/js/components/app/main.js]: +1 we need to do things like this globally to avoid inconsistency and "we will fix it later" syndrome :)

**Comment by chirimoya** [comment]
@michaelzangerle what is the current state of this PR?

**Comment by danrot** [comment]
I'm asking myself what this PR is all about at all...

**Comment by michaelzangerle** [comment]
@danrot the PR is about the handling of unsaved changes ;-) when you navigate from a form but have some changes which would be lost ...

@chirimoya the PR is nearly finished but as far as I can remember there was a special case when it did not work.

**Comment by danrot** [comment]
@michaelzangerle what should the desired behaviour be then? Should there be a warning before you change the form? Is this when changing the tabs?

**Comment by michaelzangerle** [comment]
@danrot exactly! There should be a warning which lets you decide whether to save the unsaved changes bevor navigating or discard the changes and navigate.

**Comment by danrot** [comment]
@marcelmoosbrugger Will something like this already be part of your redesign?

**Comment by webmozart** [comment]
Closing since too much time has passed. This is probably quite hard to rebase now.
